chain/chain/src/store.rs

```diff
@@ -234,6 +254,62 @@ impl ChainStore {
             }
         }
     }
+
+    pub fn check_blocks_on_same_chain(
```

Comment:
Can we simplify it and only store the head and the set of blocks, and invalidate it if the head doens't match `cur_header`?